libsolutil/Uri.h

```diff
@@ -0,0 +1,41 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <optional>
+#include <string>
+#include <string_view>
+
+namespace solidity::util
+{
+
+struct Uri
+{
+	static std::optional<Uri> parse(std::string_view _uri);
+
+	std::string scheme;
+	std::string user;
+	std::string host;
+	int port{};
```

Comment:
uint32_t?

Reply:
Ports are actually 16-bit. But why not just go with `unsigned`?

Reply:
uint16_t would be most precise; with regards to signedness, I still have that one quote from some C++ talk (by inventor(s)) in mind: "don't use unsigned unless you're doing bit-fiddling") - I don't mind either way.

Reply:
Erm - int has undefined overflow semantics, why would you want that? If you have a type that exactly matches the range then you should use it, I would say. No way to have invalid data in that case.

Reply:
Thanks for pointing out UB. I'll use uint16_t then.